"""
AIè§’è‰²æ‰®æ¼”gRPCå®¢æˆ·ç«¯ç¤ºä¾‹
æ¼”ç¤ºå¦‚ä½•é€šè¿‡gRPCè°ƒç”¨AIè§’è‰²æ‰®æ¼”åŠŸèƒ½
"""

import grpc
import sys
import os

# æ·»åŠ å½“å‰ç›®å½•åˆ°Pythonè·¯å¾„
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# å¯¼å…¥ç”Ÿæˆçš„gRPCä»£ç 
import role_play_pb2
import role_play_pb2_grpc

def run():
    # è¿æ¥åˆ°gRPCæœåŠ¡å™¨
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = role_play_pb2_grpc.RolePlayServiceStub(channel)
        
        # 1. è·å–é¢„è®¾è§’è‰²åˆ—è¡¨
        print("=== è·å–é¢„è®¾è§’è‰²åˆ—è¡¨ ===")
        try:
            response = stub.GetPresetRoles(role_play_pb2.GetPresetRolesRequest())
            for i, role in enumerate(response.roles):
                print(f"{i+1}. {role.emoji} {role.name} - {role.description}")
        except grpc.RpcError as e:
            print(f"è·å–è§’è‰²åˆ—è¡¨å¤±è´¥: {e.details()}")
            return
        
        # 2. è®¾ç½®è§’è‰²
        print("\n=== è®¾ç½®è§’è‰² ===")
        try:
            # é€‰æ‹©ç¬¬ä¸€ä¸ªè§’è‰²
            role_name = "ğŸ§™â€â™‚ï¸ æ™ºæ…§å¯¼å¸ˆ"
            response = stub.SetRole(role_play_pb2.SetRoleRequest(
                role_name=role_name
            ))
            if response.success:
                print(f"è§’è‰²è®¾ç½®æˆåŠŸ: {response.message}")
            else:
                print(f"è§’è‰²è®¾ç½®å¤±è´¥: {response.message}")
        except grpc.RpcError as e:
            print(f"è®¾ç½®è§’è‰²å¤±è´¥: {e.details()}")
            return
        
        # 3. ä¸è§’è‰²å¯¹è¯
        print("\n=== ä¸è§’è‰²å¯¹è¯ ===")
        try:
            response = stub.ChatWithRole(role_play_pb2.ChatRequest(
                message="ä½ å¥½ï¼Œèƒ½å‘Šè¯‰æˆ‘ä¸€äº›å…³äºäººå·¥æ™ºèƒ½çš„çŸ¥è¯†å—ï¼Ÿ"
            ))
            if response.success:
                print(f"AIå›å¤: {response.response}")
            else:
                print(f"å¯¹è¯å¤±è´¥: {response.error}")
        except grpc.RpcError as e:
            print(f"å¯¹è¯å¤±è´¥: {e.details()}")
            return
        
        # 4. ç»§ç»­å¯¹è¯
        print("\n=== ç»§ç»­å¯¹è¯ ===")
        try:
            response = stub.ChatWithRole(role_play_pb2.ChatRequest(
                message="é‚£æœºå™¨å­¦ä¹ å’Œæ·±åº¦å­¦ä¹ æœ‰ä»€ä¹ˆåŒºåˆ«å‘¢ï¼Ÿ"
            ))
            if response.success:
                print(f"AIå›å¤: {response.response}")
            else:
                print(f"å¯¹è¯å¤±è´¥: {response.error}")
        except grpc.RpcError as e:
            print(f"å¯¹è¯å¤±è´¥: {e.details()}")
            return
        
        # 5. é‡ç½®å¯¹è¯
        print("\n=== é‡ç½®å¯¹è¯ ===")
        try:
            response = stub.ResetConversation(role_play_pb2.ResetRequest())
            if response.success:
                print(f"é‡ç½®æˆåŠŸ: {response.message}")
            else:
                print(f"é‡ç½®å¤±è´¥: {response.message}")
        except grpc.RpcError as e:
            print(f"é‡ç½®å¯¹è¯å¤±è´¥: {e.details()}")
            return

if __name__ == '__main__':
    run()